include/aws/common/linked_list.h

```diff
@@ -19,82 +19,105 @@
-    type value;                                                                                                        \
-    struct aws_linked_list_node list;                                                                                  \
-}                                                                                                                      \
-
+/**
+ * from a pointer and a type of the struct containing the node
+ * this will get you back to the pointer of the object. member is the name of
+ * the instance of struct aws_linked_list_node in your struct.
+ */
 #define aws_container_of(ptr, type, member)  \
     ((type *)((uint8_t *)(ptr) - offsetof(type, member)))
 
-#define aws_linked_list_push_back(head, new_node) \
-    aws_linked_list_insert_before((head), (new_node))
-
-#define aws_linked_list_push_front(head, new_node) \
-    aws_linked_list_insert_after((head), (new_node))
-
-#define aws_linked_list_pop_back(head) \
-    aws_linked_list_remove((head)->prev)
-
-#define aws_linked_list_pop_front(head) \
-    head = aws_linked_list_remove(head);
-
-#define aws_linked_list_insert_before(head, new_node) \
-    aws_linked_list_insert_after(head->prev, new_node) 
-
 struct aws_linked_list_node {
     struct aws_linked_list_node *next;
     struct aws_linked_list_node *prev;
 };
 
+struct aws_linked_list {
+    struct aws_linked_list_node head;
+    struct aws_linked_list_node tail;
+};
+
 /**
- * Initializes a circular doubly-linked list.
+ * Initializes the list. List will be empty after this call.
  */
-static inline void aws_linked_list_init(struct aws_linked_list_node *head) {
-    head->next = head;
-    head->prev = head;
+static inline void aws_linked_list_init(struct aws_linked_list *list) {
+    list->head.next = &list->tail;
+    list->head.prev = NULL;
+    list->tail.prev = &list->head;
+    list->tail.next = NULL;
 }
 
 /**
- * Adds new_node between prev and next.
+ * Tests if the list is empty.
  */
-static inline void aws_linked_list_add(struct aws_linked_list_node *new_node,
-        struct aws_linked_list_node *prev,
-        struct aws_linked_list_node *next) {
+static inline int aws_linked_list_empty(struct aws_linked_list *list) {
+    return list->head.next == &list->tail;
+}
 
-    next->prev = new_node;
-    new_node->next = next;
-    new_node->prev = prev;
-    prev->next = new_node;
+/**
+ * Removes the specified node from the list (prev/next point to each other) and returns the next node in the list.
+ * If the list is empty, this function returns NULL.
+ */
+static inline void aws_linked_list_remove(struct aws_linked_list *list, struct aws_linked_list_node *node) {
+    assert(!aws_linked_list_empty(list));
+    node->prev->next = node->next;
+    node->next->prev = node->prev;
 }
 
 /**
- * Adds a node after the specified head.
+ * Append new_node.
  */
-static inline void aws_linked_list_insert_after(struct aws_linked_list_node *head, struct aws_linked_list_node *new_node) {
```

Comment:
There are no more functions for inserting to the middle of the list. Consider adding:

void aws_linked_list_insert_before(struct aws_linked_list_node *node, aws_linked_list *new_node);
void aws_linked_list_insert_after(struct aws_linked_list_node *node, aws_linked_list *new_node);